consensus/blake3pow/consensus.go

```diff
@@ -643,31 +648,282 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
+			}
+			// Reduce key size to 9 bytes and cut off the prefix
+			key = key[len(rawdb.UtxoPrefix) : rawdb.PrunedUtxoKeyWithDenominationLength+len(rawdb.UtxoPrefix)]
+			utxosCreated[i] = key
+		}
+	}
+
+	logger.Infof("UTXOs created in block %d: %d", blockHeight, len(utxosCreated))
+	if len(collidingKeys) > 0 {
+		logger.Infof("Colliding keys: %d", len(collidingKeys))
+		utxosCreated = append(utxosCreated, collidingKeys...)
+		sort.Slice(utxosCreated, func(i, j int) bool {
+			return utxosCreated[i][len(utxosCreated[i])-1] < utxosCreated[j][len(utxosCreated[j])-1]
+		})
+	}
+	newCollisions := make([][]byte, 0)
+	duplicateKeys := make(map[[36]byte]bool) // cannot use rawdb.UtxoKeyLength for map as it's not const
+	// Start by grabbing all the UTXOs created in the block (that are still in the UTXO set)
+	for _, key := range utxosCreated {
+		if len(key) != rawdb.PrunedUtxoKeyWithDenominationLength {
+			continue
+		}
+		if checkDenom {
+			if key[len(key)-1] != denomination {
+				if key[len(key)-1] > denomination {
+					break // The keys are stored in order of denomination, so we can stop checking here
+				} else {
+					continue
+				}
+			} else {
+				key = append(rawdb.UtxoPrefix, key...) // prepend the db prefix
+				key = key[:len(key)-1]                 // remove the denomination byte
+			}
+		}
+		// Check key in database
+		i := 0
+		it := chain.Database().NewIterator(key, nil)
+		for it.Next() {
+			data := it.Value()
+			if len(data) == 0 {
+				logger.Infof("Empty key found, denomination: %d", denomination)
+				continue
+			}
+			// Check if the key is a duplicate
+			if len(it.Key()) == rawdb.UtxoKeyLength {
+				key36 := [36]byte(it.Key())
+				if duplicateKeys[key36] {
+					continue
+				} else {
+					duplicateKeys[key36] = true
+				}
+			}
+			utxoProto := new(types.ProtoTxOut)
+			if err := proto.Unmarshal(data, utxoProto); err != nil {
+				logger.Errorf("Failed to unmarshal ProtoTxOut: %+v data: %+v key: %+v", err, data, key)
+				continue
+			}
+
+			utxo := new(types.UtxoEntry)
+			if err := utxo.ProtoDecode(utxoProto); err != nil {
+				logger.WithFields(log.Fields{
+					"key":  key,
+					"data": data,
+					"err":  err,
+				}).Error("Invalid utxo Proto")
+				continue
+			}
+			if checkDenom && utxo.Denomination != denomination {
+				continue
+			}
+			txHash, index, err := rawdb.ReverseUtxoKey(it.Key())
+			if err != nil {
+				logger.WithField("err", err).Error("Failed to parse utxo key")
+				continue
+			}
+			lock.Lock()
+			*utxosDelete = append(*utxosDelete, types.UTXOHash(txHash, index, utxo))
+			if deleteFromDb {
+				batch.Delete(it.Key())
+				*trimmedUtxos = append(*trimmedUtxos, &types.SpentUtxoEntry{OutPoint: types.OutPoint{txHash, index}, UtxoEntry: utxo})
+			}
+			*utxoSetSize--
+			lock.Unlock()
+			i++
+			if i >= types.MaxTrimCollisionsPerKeyPerBlock {
```

Comment:
What is the rationale for condensing the keys like this? why not use the full size? Asking this because want to know if doing that takes away all this complexity

Reply:
The key is half the UTXO size. To support 1 billion UTXOs the key must be as condensed as possible. Otherwise we can’t store the UTXO set and the trimming data together, it’s too much data. However currently we don’t prune at all, so most of this code path isn't executed.